petastorm/pytorch.py

```diff
@@ -348,7 +365,38 @@ def _instantiate_noop_buffer():
 
         return shuffling_buffer, other_shuffling_buffer, instantiate_buffer_fn
 
+    def _worker(self):
+        for b in self._iter_impl_worker():
+            if self._finished_event.is_set():
+                break
+            self._async_batch_queue.put(b)
+        else:
+            self._async_batch_queue.put(None)
+
     def _iter_impl(self):
+        if self._async_batch_queue:
+            # Get shuffled batches from async shuffle thread
+            if not self._started:
+                # If reuse dataloader in a new pass, async shuffle thread has to be relaunched.
+                # It is finished in previous pass when reaching end of data.
```

Comment:
Is the restart scenario covered by a unit test?

Reply:
This is covered by using multiple epochs in the unit test, for example:
```
for _ in range(num_epochs):
            # Check if dataloder works correctly with multiple passes.
            _check_simple_reader(loader, synthetic_dataset.data, TORCH_BATCHABLE_FIELDS - {TestSchema.sensor_name})
```

The 2nd and 3rd epoch will hit the restarting case, because previous round of data read already completes.